### PR TITLE
[WOR-129] Add model updates to support workspace type

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
In #1618  we added the `workspace_type` field to the `WORKSPACE` table. The next step is to thread the new field through to the related model classes.

## This PR
* Adds a new `WorkspaceType` enum to capture states this field can take
* Threads this field through the `Workspace` model and associated calls sites
    * This field is being hardcoded to `rawls` to start. The notion is that Azure workspaces created via the updated `createWorkspace` endpoint will receive a `mc` value.
* **Note** This PR is being created against #1618. I expect this will merge to that PR and then that one will go out as a single unit.

____

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 1]
```diff
@@ -329,6 +331,35 @@ object WorkspaceShardStates {
...
+  /**
+    * Represents the system-of-record for a workspace
+    * rawls workspaces are owned and managed by rawls
+    * mc workspaces are owned and managed by workspace manager
+    */
+  sealed trait WorkspaceType extends RawlsEnumeration[WorkspaceType] {
+    override def toString: String = WorkspaceType.toString(this)
+    override def withName(name: String): WorkspaceType = WorkspaceType.withName(name)
+  }
+
```
Comment (cahrens) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:344]: @aherbst-broad is it feasible to add test coverage for these lines?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 2]
```diff
@@ -481,15 +484,15 @@ trait WorkspaceComponent {
...
+      WorkspaceRecord(workspace.namespace, workspace.name, UUID.fromString(workspace.workspaceId), workspace.bucketName, workspace.workflowCollectionName, new Timestamp(workspace.createdDate.getMillis), new Timestamp(workspace.lastModified.getMillis), workspace.createdBy, workspace.isLocked, 0, workspace.workspaceVersion.value, workspace.googleProjectId.value, workspace.googleProjectNumber.map(_.value), workspace.currentBillingAccountOnGoogleProject.map(_.value), workspace.billingAccountErrorMessage, workspace.completedCloneWorkspaceFileTransfer.map(dateTime => new Timestamp(dateTime.getMillis)), shardState = WorkspaceShardStates.Sharded.toString, workspaceType = workspace.workspaceType.toString)
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:487]: Nit: these lines exceptionally long and unreadable (to me).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -2231,7 +2232,7 @@ class WorkspaceService(protected val userInfo: UserInfo,
                                          googleProjectNumber: Option[GoogleProjectNumber],
                                          currentBillingAccountOnWorkspace: Option[RawlsBillingAccountName],
                                          dataAccess: DataAccess,
-                                         parentSpan: Span = null): ReadWriteAction[Workspace] = {
+                                         parentSpan: Span = null, workspaceType: WorkspaceType = WorkspaceType.RawlsWorkspace): ReadWriteAction[Workspace] = {
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Nit: it looks like the style here is one argument per line (which I like for readability).

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 4]
```diff
@@ -329,6 +331,35 @@ object WorkspaceShardStates {
...
+  def withName(name: String): WorkspaceType = name.toLowerCase match {
+    case "rawls" => RawlsWorkspace
+    case "mc" => McWorkspace
```
Comment (cahrens) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:347]: What does "mc" stand for?

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 5]
```diff
@@ -939,7 +973,9 @@ class WorkspaceJsonSupport extends JsonSupport {
 
   implicit val WorkspaceShardStateFormat = rawlsEnumerationFormat(WorkspaceShardStates.withName)
 
-  implicit val WorkspaceDetailsFormat = jsonFormat18(WorkspaceDetails.applyOmitShardState)
+  implicit val WorkspaceStageFormat = rawlsEnumerationFormat(WorkspaceType.withName)
```
Comment (cahrens) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: `WorkspaceStageFormat` seems like an odd name for this.

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 1 above)
Comment (aherbst-broad) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:344]: Good call, [done](https://github.com/broadinstitute/rawls/pull/1629/commits/33787a8f402e4acb36c6730a975af7fbafc156c0)

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 4 above)
Comment (aherbst-broad) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:347]: "Multi-cloud", this is an attempt for a bit of consistency with WSM's naming scheme [here](https://github.com/DataBiosphere/terra-workspace-manager/blob/540521667f235a6d2f507944683763912a6c3148/service/src/main/java/bio/terra/workspace/service/workspace/model/WorkspaceStage.java#L12-L13).

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 5 above)
Comment (aherbst-broad) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: Ah, good eye, this was an artifact of a previous naming scheme, fixed.

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 4 above)
Comment (cahrens) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:347]: I see. Could you add a comment here that "mc" stands for "multi-cloud"? Somehow I feel better about abbreviations when I know what they stand for. :)

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 6]
```diff
@@ -329,6 +331,35 @@ object WorkspaceShardStates {
   case object Unknown extends WorkspaceShardState
 }
 
+object WorkspaceType {
+  /**
+    * Represents the system-of-record for a workspace
+    * rawls workspaces are owned and managed by rawls
+    * mc workspaces are owned and managed by workspace manager
```
Comment (cahrens) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]:
```suggestion
    * Represents the system-of-record for a workspace.
    * `rawls` workspaces are owned and managed by rawls.
    * `mc` (multi-cloud) workspaces are owned and managed by workspace manager.
```

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 4 above)
Comment (cahrens) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:347]: I made a suggestion in the documentation block.

**Review by cahrens — APPROVED** [review]
LGTM, though I would appreciate documenting "mc".

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 7]
```diff
@@ -329,6 +331,35 @@ object WorkspaceShardStates {
...
+  def toString(wt: WorkspaceType): String = {
+    wt match {
+      case RawlsWorkspace => "rawls"
+      case McWorkspace => "mc"
+      case _ => throw new RawlsException(s"Invalid WorkspaceType [${wt}]")
+    }
+  }
+
+  case object RawlsWorkspace extends WorkspaceType
+  case object McWorkspace extends WorkspaceType
```
Comment (marctalbott) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:360]: Where do Rawls workspaces that have stubs in WSM for use with TDR fit into this? Are they still `RawlsWorkspace`s?

**Review by marctalbott — APPROVED** [review]

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 7 above)
Comment (aherbst-broad) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:360]: @marctalbott yes, for the time being. There will have to be a migration effort in the future to get those to the "mc" world, but this is purely for azure workspaces right now.

**Pull request merged**